packages/fast-usdc-contract/src/utils/settlement-matching.ts

```diff
@@ -0,0 +1,128 @@
+ *
+ * Implementation includes memoization and depth-limiting to ensure performance even with
+ * larger transaction sets, though typical match depth is 2-3 transactions.
+ */
+
+import type { MapStore } from '@agoric/swingset-liveslots';
+import type { NobleAddress, PendingTx } from '@agoric/fast-usdc/src/types.ts';
+import { appendToSortedStoredArray } from './store.ts';
+
+/**
+ * Max number of txs to settle against a single mint.
+ * Much higher than observed value of 2-3.
+ */
+const MAX_MATCH_DEPTH = 25;
+
+/** bigint‑safe descending comparator (largest first) */
+const comparePendingTxDesc = (a: PendingTx, b: PendingTx): number => {
+  if (a.tx.amount < b.tx.amount) return 1;
+  if (a.tx.amount > b.tx.amount) return -1;
+  return 0;
+};
+
+harden(comparePendingTxDesc);
+
+/**
+ * Depth‑first, memoised, largest‑first chooser.
+ * Returns the *first* exact combination or [] if none.
+ */
+const greedyMatch = (
+  pending: readonly PendingTx[],
+  target: bigint,
+): PendingTx[] => {
+  if (target === 0n) return [];
+  if (pending.length === 0) return [];
+
+  const memo = new Map<string, boolean>();
+  const path: PendingTx[] = [];
+
+  const dfs = (index: number, sum: bigint): boolean => {
+    if (sum === target) return true;
+    if (sum > target) return false; // unreachable with current continue‑on‑overshoot logic
+    if (path.length >= MAX_MATCH_DEPTH) return false;
+
+    if (index >= pending.length) return false;
+
+    const key = `${index}:${sum}`;
+    if (memo.get(key) === false) return false; // skip known dead end
+
+    for (let i = index; i < pending.length; i += 1) {
+      const tx = pending[i];
+      const nextSum = sum + tx.tx.amount;
+      if (nextSum > target) continue; // overshoot, skip
+
+      path.push(tx);
+      if (dfs(i + 1, nextSum)) return true;
+      path.pop();
+    }
+
+    memo.set(key, false);
+    return false;
+  };
+
+  return dfs(0, 0n) ? harden([...path]) : [];
+};
+
+harden(greedyMatch);
+
+export const makeSettlementMatching = (
+  pendingSettleTxs: MapStore<NobleAddress, PendingTx[]>,
+) => {
+  /** add to per‑address queue, preserving descending sort */
+  const addPendingSettleTx = (pending: PendingTx): void => {
+    appendToSortedStoredArray(
+      pendingSettleTxs,
+      pending.tx.forwardingAddress,
+      pending,
+      comparePendingTxDesc,
+    );
+  };
+
+  /**
+   * Attempt to satisfy `amount` for `nfa`.  Returns the matched txs (possibly
+   * empty).  Updates the MapStore so those txs are no longer pending.
+   */
+  const matchAndDequeueSettlement = (
+    nfa: NobleAddress,
+    amount: bigint,
+  ): PendingTx[] => {
+    if (!pendingSettleTxs.has(nfa)) return [];
+    const list = pendingSettleTxs.get(nfa);
+
+    // 1. exact
+    const ix = list.findIndex(tx => tx.tx.amount === amount);
+    if (ix >= 0) {
+      const match = list[ix];
+      const remaining = [...list.slice(0, ix), ...list.slice(ix + 1)];
+      remaining.length
```

Comment:
please no unused expressions
```suggestion
      if (remaining.length) {
```
same below